libs/wire-api-federation/test/Test/Wire/API/Federation/ClientSpec.hs

```diff
@@ -0,0 +1,148 @@
+
+outwardService :: IO OutwardResponse -> SingleServerT info Outward (MockT ServerErrorIO) _
+outwardService res = Mu.singleService (Mu.method @"call" (callOutward res))
+
+callOutward :: (MonadServer m, MonadState RecievedRequests m) => IO OutwardResponse -> FederatedRequest -> m OutwardResponse
+callOutward res req = do
+  modify (<> [req])
+  liftIO res
+
+mkSuccessResponse :: Aeson.ToJSON a => a -> IO OutwardResponse
+mkSuccessResponse = pure . OutwardResponseBody . LBS.toStrict . Aeson.encode
+
+mkErrorResponse :: OutwardError -> IO OutwardResponse
+mkErrorResponse = pure . OutwardResponseError
+
+-- This is mostly copy-pasta from brig-integration. Perhaps this should be part
+-- of the wire-api-federation libarary?
+withMockFederator :: IO OutwardResponse -> FederatorClient 'Brig (ExceptT FederationClientError IO) a -> IO (Either FederationClientError a, RecievedRequests)
+withMockFederator res action = do
+  let port = 47282 -- TODO: Make this actually arbitrary?
+  requestsRef <- newIORef []
+  federatorThread <- Async.async $ runGRpcAppTrans msgProtoBuf port (runMockT requestsRef) (outwardService res)
+  -- FUTUREWORK: This takes exactly 2 seconds, perhaps something sleeps when grpc server starts, investigate
+  serverStarted <- retryWhileN 5 not (isPortOpen port)
+  serverStarted `shouldBe` True
+
+  let cfg = grpcClientConfigSimple "127.0.0.1" (fromIntegral port) False
+  client <- assertRight =<< createGrpcClient cfg
+
+  actualResponse <-
+    runExceptT (runFederatorClientWith client targetDomain originDomain action)
+      `finally` Async.cancel federatorThread
+  reqs <- readIORef requestsRef
+  pure (actualResponse, reqs)
+  where
+    targetDomain = Domain "target.example.com"
+    originDomain = Domain "origin.example.com"
+    isPortOpen :: Int -> IO Bool
+    isPortOpen port = do
+      let sockAddr = SockAddrInet (fromIntegral port) (tupleToHostAddress (127, 0, 0, 1))
+      bracket (socket AF_INET Stream 6 {- TCP -}) close' $ \sock -> do
+        portRes <- try $ connect sock sockAddr
+        case portRes of
+          Right () -> return True
+          Left e ->
+            if (Errno <$> ioe_errno e) == Just eCONNREFUSED
+              then return False
+              else throwIO e
+
+newtype MockT m a = MockT {unMock :: ReaderT (IORef RecievedRequests) m a}
+  deriving newtype (Functor, Applicative, Monad, MonadReader (IORef RecievedRequests), MonadIO)
+
+instance (MonadError ServerError m) => MonadError ServerError (MockT m) where
+  throwError err = MockT . lift $ throwError err
+  catchError action f = do
+    r <- ask
+    MockT . lift $ catchError (runMockT r action) (runMockT r . f)
+
+instance MonadIO m => MonadState RecievedRequests (MockT m) where
+  get = readIORef =<< ask
+  put x = do
+    ref <- ask
+    writeIORef ref x
+
+runMockT :: IORef RecievedRequests -> MockT m a -> m a
+runMockT ref mock = runReaderT (unMock mock) ref
+
+retryWhileN :: forall a m. (MonadIO m) => Int -> (a -> Bool) -> m a -> m a
```

Comment:
It sounds like retryWhileN should really move to a lower-level library and not being re-defined again, I think there's already a range of copies in the codebase:

```
ag "retryWhileN ::"
services/gundeck/test/integration/API.hs
936:retryWhileN :: (MonadIO m) => Int -> (a -> Bool) -> m a -> m a

services/brig/test/integration/Util.hs
750:retryWhileN :: (MonadIO m) => Int -> (a -> Bool) -> m a -> m a

services/galley/test/integration/API/Util.hs
1353:retryWhileN :: (MonadIO m) => Int -> (a -> Bool) -> m a -> m a

libs/cassandra-util/src/Cassandra/Schema.hs
251:retryWhileN :: (MonadIO m) => Int -> (a -> Bool) -> m a -> m a
```

Reply:
I couldn't figure out where it should go, then I thought this type of duplication in test code doesn't really hurt us, so I gave up :shrug: